### PR TITLE
[Model Enableing]Phi3-4k&Phi3-128k

### PR DESCRIPTION
**Pull request by intellinjun**
## Type of Change

[Model Enableing]
Jira

- [Phi3 enabling for neural speed](https://jira.devtools.intel.com/browse/NLPTOOLKIU-1368)
## Description

Enable Phi3
detail description 

- [phi3-4k](https://huggingface.co/microsoft/Phi-3-mini-4k-instruct/tree/main)
- [phi3-128k](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct)
- longrope

### CODE REVIEW

**Comment by intellinjun** [comment]
image
128k  q4j32 result

**Comment by intellinjun** [comment]
image
128k q4j32 boolq acc

File: neural_speed/convert/convert_phi3.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    fout.write(struct.pack("i", 0))  # do_layer_norm_before (for opt)
+
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))  # n_experts
+    fout.write(struct.pack("i", 0))  # n_expert_used
+    fout.write(struct.pack("i", 0)) # n_embd_head_k for gemma
+    fout.write(struct.pack("f", hparams.get("layer_norm_eps", 1e-5)))  # rms_norm_eps or layer_norm_eps
+    fout.write(struct.pack("f", 10000.0))  # freq_base
```
Comment (a32543254) [inline neural_speed/convert/convert_phi3.py:93]: could freq_base read from config.json ?
like:
hparams.get("freq_base ", 1e-5)

File: neural_speed/core/ne_layers.c [hunk 2]
```diff
@@ -9040,6 +9047,38 @@ static void ne_compute_forward_rope_f32(const struct ne_compute_params* params,
             dst_data[n_dims] = x2 * cos_block_theta - x3 * sin_block_theta;
             dst_data[n_dims / 2 * 3] = x2 * sin_block_theta + x3 * cos_block_theta;
           }
+        } else if (is_longrope) {
+          // TODO: this is probably wrong, but I can't figure it out ..
```
Comment (a32543254) [inline neural_speed/core/ne_layers.c:9051]: this is probably wrong, but I can't figure it out
what is the mean of this ？

File: neural_speed/core/ne_layers.c [hunk 3]
```diff
@@ -9203,7 +9242,31 @@ static void ne_compute_forward_rope_f16(const struct ne_compute_params* params,
 
         float theta = freq_scale * (float)p;
 
-        if (!is_neox) {
+        if (is_longrope) {
+          float scale_factor = 1.1902380714238083;
```
Comment (a32543254) [inline neural_speed/core/ne_layers.c:9246]: this value is from paper or could also get from config ？

**Review by a32543254 — APPROVED** [review]
LGTM

File: neural_speed/core/ne_layers.c (hunk 2 above)
Comment (intellinjun) [reply neural_speed/core/ne_layers.c]: removed

File: neural_speed/core/ne_layers.c (hunk 3 above)
Comment (intellinjun) [reply neural_speed/core/ne_layers.c]: get from max_seq_len

**Review by Zhenzhong1 — APPROVED** [review]

**Comment by intellinjun** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


Model | Input | Output | Batchsize | Cores/Instance | Precision | Eval Time per   Token | 1st Latency | Total Time
-- | -- | -- | -- | -- | -- | -- | -- | --
phi3 | 32 | 32 | 1 | 32 | q4_j_i8_g128 | 16.63 | 76.55 | 592.08
phi3 | 1024 | 32 | 1 | 32 | q4_j_i8_g128 | 18.43 | 1965.66 | 2536.99
phi3 | 2012 | 32 | 1 | 32 | q4_j_i8_g128 | 20.34 | 3896.75 | 4527.29
phi3 | 32 | 32 | 1 | 32 | q4_j_i8_g32 | 19.11 | 88.11 | 680.52
phi3 | 1024 | 32 | 1 | 32 | q4_j_i8_g32 | 23.11 | 2590.1 | 3306.51
phi3 | 2012 | 32 | 1 | 32 | q4_j_i8_g32 | 25.11 | 5244.91 | 6023.32
phi3 | 32 | 32 | 1 | 32 | q4_0 | 21.7 | 264.01 | 936.71
phi3 | 1024 | 32 | 1 | 32 | q4_0 | 25.68 | 8516.57 | 9312.65
phi3 | 2012 | 32 | 1 | 32 | q4_0 | 26.77 | 16492.27 | 17322.14



</body>

</html>

**Pull request merged**